Sources/Subscription/Managers/SubscriptionManagerV2.swift

```diff
@@ -0,0 +1,423 @@
+        let serviceResponse = try await subscriptionEndpointService.getCustomerPortalURL(accessToken: tokenContainer.accessToken, externalID: tokenContainer.decodedAccessToken.externalID)
+        guard let url = URL(string: serviceResponse.customerPortalUrl) else {
+            throw SubscriptionEndpointServiceError.noData
+        }
+        return url
+    }
+
+    // MARK: - User
+    public var isUserAuthenticated: Bool {
+        oAuthClient.isUserAuthenticated
+    }
+
+    public var userEmail: String? {
+        return oAuthClient.currentTokenContainer?.decodedAccessToken.email
+    }
+
+    // MARK: -
+
+    @discardableResult public func getTokenContainer(policy: AuthTokensCachePolicy) async throws -> TokenContainer {
+        do {
+            Logger.subscription.debug("Get tokens \(policy.description, privacy: .public)")
+
+            let referenceCachedTokenContainer = try? await oAuthClient.getTokens(policy: .local) // the currently stored one
+            let referenceCachedEntitlements = referenceCachedTokenContainer?.decodedAccessToken.subscriptionEntitlements
+            let resultTokenContainer = try await oAuthClient.getTokens(policy: policy)
+            let newEntitlements = resultTokenContainer.decodedAccessToken.subscriptionEntitlements
+
+            // Send notification when entitlements change
+            if referenceCachedEntitlements != newEntitlements {
+                Logger.subscription.debug("Entitlements changed - New \(newEntitlements) Old \(String(describing: referenceCachedEntitlements))")
+                NotificationCenter.default.post(name: .entitlementsDidChange, object: self, userInfo: [UserDefaultsCacheKey.subscriptionEntitlements: newEntitlements])
+            }
+
+            if referenceCachedTokenContainer == nil { // new login
+                Logger.subscription.debug("New login detected")
+                NotificationCenter.default.post(name: .accountDidSignIn, object: self, userInfo: nil)
+            }
+            return resultTokenContainer
+        } catch OAuthClientError.refreshTokenExpired {
+            return try await throwAppropriateDeadTokenError()
+        } catch {
+            throw SubscriptionManagerError.tokenUnavailable(error: error)
+        }
+    }
+
+    /// If the client succeeds in making a refresh request but does not get the response, then the second refresh request will fail with `invalidTokenRequest` and the stored token will become unusable and un-refreshable.
+    private func throwAppropriateDeadTokenError() async throws -> TokenContainer {
+        Logger.subscription.fault("Dead token detected")
+        do {
+            let subscription = try await subscriptionEndpointService.getSubscription(accessToken: "", // Token is unused
+                                                                                     cachePolicy: .returnCacheDataDontLoad)
+            switch subscription.platform {
```

Comment:
I suspect it is highly unlikely this logic to succeed, if the token is dead it means that it is long time since the app interacted with BE, hence the subscription cache will be stale (currently it is set to 20 minutes).

Reply:
I agree. The dead token handling needs to be reviewed completely as you pointed out in your first PR review, and I created a task dedicated to that https://app.asana.com/0/72649045549333/1209252725803000